### PR TITLE
Enabling DPC++ for SYCL-BLAS.

### PR DESCRIPTION
**Pull request by mehdi-goli**

### CODE REVIEW

File: src/executors/kernel_constructor.hpp [hunk 1]
```diff
@@ -86,6 +86,7 @@ Specialised case for using_local_memory == subgroup, which contains a subgroup
 local accessor.
 @tparam value_t Value type of the accessor.
 */
+#if 0
```
Comment (ProGTX) [inline src/executors/kernel_constructor.hpp:89]: Marking this so it doesn't get forgotten

File: src/executors/kernel_constructor.hpp [hunk 2]
```diff
@@ -213,10 +215,11 @@ struct ExpressionTreeFunctor {
...
                                          expression_tree_t t)
       : scratch_(scratch), t_(t) {}
-  SYCL_BLAS_INLINE void operator()(cl::sycl::nd_item<1> i) {
-    t_.adjust_access_displacement();
+  SYCL_BLAS_INLINE void operator()(cl::sycl::nd_item<1> i) const {
+    expression_tree_t& non_const_t=*const_cast<expression_tree_t*>(&t_);
+    non_const_t.adjust_access_displacement();
     ExpressionTreeEvaluator<using_local_memory, expression_tree_t,
-                            value_t>::eval(t_, scratch_, i);
+                            value_t>::eval(non_const_t, scratch_, i);
```
Comment (ProGTX) [inline src/executors/kernel_constructor.hpp:222]: I guess this is more of a long term question, but can the `eval` be changed to take a const object?

File: src/operations/blas1_trees.hpp [hunk 3]
```diff
@@ -254,7 +254,7 @@ SYCL_BLAS_INLINE void DoubleAssign<lhs_1_t, lhs_2_t, rhs_1_t, rhs_2_t>::bind(
 template <typename lhs_1_t, typename lhs_2_t, typename rhs_1_t,
           typename rhs_2_t>
 SYCL_BLAS_INLINE void
-DoubleAssign<lhs_1_t, lhs_2_t, rhs_1_t, rhs_2_t>::adjust_access_displacement() {
+DoubleAssign<lhs_1_t, lhs_2_t, rhs_1_t, rhs_2_t>::adjust_access_displacement()  {
```
Comment (ProGTX) [inline src/operations/blas1_trees.hpp:257]: Some of the changes in this file indicate it's not formatted

File: src/operations/blas3/gemm_interleaved.hpp [hunk 4]
```diff
@@ -65,7 +65,7 @@ SYCL_BLAS_INLINE void store(T packet, PtrT ptr) {
 template <address_t Address = address_t::global_space, class T, int Dim,
           class PtrT>
 SYCL_BLAS_INLINE void load(cl::sycl::vec<T, Dim> &packet, PtrT ptr) {
-  packet.template load<Address>(0, ptr);
+  packet.template load<Address>(0, cl::sycl::multi_ptr<const T, Address>(ptr));
```
Comment (ProGTX) [inline src/operations/blas3/gemm_interleaved.hpp:68]: Why are these changes required? Doesn't the pointer get implicitly converted to a `multi_ptr`?

File: CMakeLists.txt [hunk 5]
```diff
@@ -53,20 +53,34 @@ list(APPEND CMAKE_MODULE_PATH
...
+    -O3
+    -fsycl-split-modules=20
+    -mllvm -inline-threshold=10000
+    -Xclang -cl-mad-enable
+    # We add some flags to workaround OpenCL platform bugs, see ComputeCpp documentation
+    -no-serial-memop
+  )
+endif()
+
+set(CMAKE_CXX_STANDARD 17)
```
[the remark this answers is not in the document]
Comment (mehdi-goli) [reply CMakeLists.txt]: Intel compiler requires c++17. Moved to intel

File: CMakeLists.txt [hunk 6]
```diff
@@ -78,19 +92,27 @@ if (MSVC)
 endif()
 
 # By default, tall and skinny Gemm is enabled (for better performance)
-option(GEMM_TALL_SKINNY_SUPPORT "Whether to enable tall and skinny Gemm" ON)
+option(GEMM_TALL_SKINNY_SUPPORT "Whether to enable tall and skinny Gemm" OFF)
```
[the remark this answers is not in the document]
Comment (mehdi-goli) [reply CMakeLists.txt]: Yes it is. It has been enabled

File: CMakeLists.txt [hunk 7]
```diff
@@ -78,19 +92,27 @@ if (MSVC)
...
 
 set(CBLAS_INCLUDE ${CMAKE_CURRENT_SOURCE_DIR}/external/cblas/include)
-set(COMPUTECPP_SDK_INCLUDE ${CMAKE_CURRENT_SOURCE_DIR}/external/computecpp-sdk/include)
+if(is_computecpp)
+  set(COMPUTECPP_SDK_INCLUDE ${CMAKE_CURRENT_SOURCE_DIR}/external/computecpp-sdk/include)
+elseif(is_dpcpp)
+  set(COMPUTECPP_SDK_INCLUDE ${CMAKE_CURRENT_SOURCE_DIR}/external/computecpp-sdk/include)
+  get_target_property(DPCPP_INCLUDE_DIRS DPCPP::DPCPP INTERFACE_INCLUDE_DIRECTORIES)
+  set(ComputeCpp_INCLUDE_DIRS ${DPCPP_INCLUDE_DIRS})
+  Message(ComputeCpp_INCLUDE_DIRS ${DPCPP_INCLUDE_DIRS})
```
[the remark this answers is not in the document]
Comment (mehdi-goli) [reply CMakeLists.txt]: Removed in the final PR

File: CMakeLists.txt [hunk 8]
```diff
@@ -78,19 +92,27 @@ if (MSVC)
...
 option(GEMM_VECTORIZATION_SUPPORT "Whether to enable vectorization in Gemm kernels" OFF)
 
 add_definitions(-DCL_TARGET_OPENCL_VERSION=220)
 
 set(CBLAS_INCLUDE ${CMAKE_CURRENT_SOURCE_DIR}/external/cblas/include)
-set(COMPUTECPP_SDK_INCLUDE ${CMAKE_CURRENT_SOURCE_DIR}/external/computecpp-sdk/include)
+if(is_computecpp)
+  set(COMPUTECPP_SDK_INCLUDE ${CMAKE_CURRENT_SOURCE_DIR}/external/computecpp-sdk/include)
+elseif(is_dpcpp)
+  set(COMPUTECPP_SDK_INCLUDE ${CMAKE_CURRENT_SOURCE_DIR}/external/computecpp-sdk/include)
```
[the remark this answers is not in the document]
Comment (mehdi-goli) [reply CMakeLists.txt]: extra one removed

File: CMakeLists.txt [hunk 9]
```diff
@@ -115,18 +137,37 @@ add_subdirectory(src)
...
-  # The sycl_blas target is just a collection of other objects so
-  # the linked libraries are not going to be propagated to this target.
-  # This requires manual linking against ComputeCpp.lib on Windows.
-  target_link_libraries(sycl_blas PRIVATE ComputeCpp::ComputeCpp)
+  if(is_computecpp)
+
+    # On Windows, all symbols must be resolved at link time for DLLs.
+    # The sycl_blas target is just a collection of other objects so
+    # the linked libraries are not going to be propagated to this target.
+    # This requires manual linking against ComputeCpp.lib on Windows.
```
[the remark this answers is not in the document]
Comment (mehdi-goli) [reply CMakeLists.txt]: comment has been generalised

File: src/executors/kernel_constructor.hpp (hunk 1 above)
Comment (mehdi-goli) [reply src/executors/kernel_constructor.hpp]: Fixed

File: src/executors/kernel_constructor.hpp (hunk 2 above)
Comment (mehdi-goli) [reply src/executors/kernel_constructor.hpp:222]: Discussed internally

File: src/operations/blas1_trees.hpp (hunk 3 above)
Comment (mehdi-goli) [reply src/operations/blas1_trees.hpp]: Clang-format appliled

File: src/operations/blas3/gemm_interleaved.hpp (hunk 4 above)
Comment (mehdi-goli) [reply src/operations/blas3/gemm_interleaved.hpp:68]: Without this, it gives ambiguous error in intel DPC++

**Comment by mehdi-goli** [comment]
@mcleary It would be good if you could check the windows pass if possible

**Comment by mcleary** [comment]
I'm testing this on Windows, I should have an answer if anything needs to be changed soon

File: src/executors/kernel_constructor.hpp (hunk 2 above)
Comment (ProGTX) [reply src/executors/kernel_constructor.hpp:222]: The answer is complicated, but probably not :)

File: src/operations/blas3/gemm_interleaved.hpp (hunk 4 above)
Comment (ProGTX) [reply src/operations/blas3/gemm_interleaved.hpp:68]: It's actually correct SYCL 1.2.1 to use a `multi_ptr`, so this change is required, thanks

**Review by ProGTX — APPROVED** [review]

**Comment by mcleary** [comment]
> I'm testing this on Windows, I should have an answer if anything needs to be changed soon

@mehdi-goli unfortunately the dpcpp integration on Windows is not working out-of-the-box, but this is because the general support for dpcpp using CMake on Windows is not good at the moment. I would say for you to merge this and leave the work on windows for another pull request

**Comment by mehdi-goli** [comment]
> > I'm testing this on Windows, I should have an answer if anything needs to be changed soon
> 
> @mehdi-goli unfortunately the dpcpp integration on Windows is not working out-of-the-box, but this is because the general support for dpcpp using CMake on Windows is not good at the moment. I would say for you to merge this and leave the work on windows for another pull request

Hi Thales, Thanks for checking, that what I was guessing, unfortunately. So for now let's only use it for Linux and we keep the windows support for later as a separate PR when DPC++ is ready for it.  Let me know if you happy for the PR to be merged as is

**Review by Bensuo — APPROVED** [review]
Looks good to me!
